clustering/zookeeper.py

```diff
@@ -0,0 +1,279 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = """
+---
+module: zookeeper
+short_description: Manipulate ZooKeeper keys
+description:
+  - Create and modify I(ZooKeeper) keys.
+requirements:
+  - C(kazoo)
+author: James Bowes
+options:
+  name:
+    description:
+      - The key to manipulate.
+    aliases: [key, path]
+    required: true
+  value:
+    description:
+      - An optional value to set for the key.
+  state:
+    description:
+      - Indicates the desired key state C(present) indicates the key should
+        exist with the given value, or be empty if no value is provided.
+        C(absent) indicates the path should not exist.
+    default: present
+    choices: ["present", "absent"]
+  acls:
+    description:
+      - A single ACL, or list of ACLs for the node, of the form
+        'scheme:identity:perms'.
+      - Valid schemes are 'world', 'digest', 'userpass', 'ip'.
+      - Valid perms are 'c', 'r', 'w', 'd', 'a'.
+  exclusive:
+    description:
+      - Whether to remove all other non-specified ACLs from the key.
+    default: no
+    choices: ["yes", "no"]
+  login_host:
+    description:
+      - The host running I(ZooKeeper).
+    default: localhost
+  login_port:
+    description:
+      - The port to connect to.
+    default: 2181
+  login_user:
+    description:
+      - Optional username for authentication.
+  login_password:
+    description:
+      - Optional password for authentication.
+"""
+
+EXAMPLES = """
+# Ensure a key exists and is empty
+- zookeeper: name=/my/key
+
+# Ensure a key exists with a value
+- zookeeper: name=/my/other/key value="has a nice value"
+
+# Ensure a key does not exist
+- zookeeper: name=/make/sure/this/is/gone state=absent
+
+# Allow user1 to perform all actions on a key
+- zookeeper:
+    name: /key
+    value: "something"
+    acls:
+      - "userpass:user1:password:crwda"
+
+# Grant admin permissions on a key to user1.
+# Allow user2 to create and delete child keys.
+# Allow everyone to read the key and its children.
+- zookeeper:
+    name: /key
+    value: "something"
+    acls:
+      - "digest:user1:bYeZNFxkmtQrhcuNivU587R4lRg=:a"
+      - "userpass:user2:password:cd"
+      - "world:anyone:r"
+"""
+
+import sys
+
+# Prevent kazoo logging from messing up our output.
+import logging
+logging.basicConfig()
+logging.getLogger("kazoo").propagate = False
+
+try:
+    from kazoo.client import KazooClient
+    from kazoo.exceptions import NoAuthError, NoNodeError, ZookeeperError
+    from kazoo.security import make_acl, make_digest_acl_credential
+except ImportError:
+    print "failed=True msg='kazoo is required for this module. "\
+          "see https://kazoo.readthedocs.org/en/latest/install.html'"
+    sys.exit(1)
+
+# Initialiize Ansible Module
+# import module snippets
+from ansible.module_utils.basic import *
```

Comment:
you might want to do this at the bottom of the file, this is not a real import and will mess up line numbers for error reporting